### PR TITLE
Expose more configurations for CA

### PR DESCRIPTION
**Pull request by prashanth26**
**How to categorize this PR?**

/area auto-scaling
/kind enhancement

**What this PR does / why we need it**:
The cluster autoscaler for a shoot can be configured with more parameters.
  - expander lets you configure the algorithm during scale-up. (Default: least-waste)
  - maxNodeProvisionTime is the maximum provisioning time before declaring the worker pool is unhealthy by CA. (Default: 20m)
  - Documentation has also been updated to reflect the correct default values.
  - Aligning the default `max-node-provisioning-time` with `machine creation-timeout`.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
This allows more configurability for end-users at the cluster Autoscaler

**Release note**:

```feature user
Cluster Autoscaler is now more configurable at the shoot level with options like `expander` and `maxNodeProvisionTime`. The default value for `expander` is `least-waste` and for `maxNodeProvisionTime` it is `20m`. This can be overridden by flags on the `shoot.spec.kubernetes.clusterAutoscaler` section. 
```

### CODE REVIEW

File: pkg/apis/core/validation/shoot.go [hunk 1]
```diff
@@ -793,6 +793,10 @@ func ValidateClusterAutoscaler(autoScaler core.ClusterAutoscaler, fldPath *field
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("scaleDownUtilizationThreshold"), *threshold, "can not be greater than 1.0"))
 		}
 	}
+	if maxNodeProvisionTime := autoScaler.MaxNodeProvisionTime; maxNodeProvisionTime != nil && maxNodeProvisionTime.Duration < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("maxNodeProvisionTime"), *maxNodeProvisionTime, "can not be negative"))
+	}
+
```
Comment (vpnachev) [inline pkg/apis/core/validation/shoot.go:805]: It will be nice the add a validation also for the `expander` field.

File: pkg/apis/core/v1beta1/defaults.go [hunk 2]
```diff
@@ -321,6 +321,34 @@ func SetDefaults_Worker(obj *Worker) {
 	}
 }
 
+// SetDefaults_ClusterAutoscaler sets default values for ClusterAutoscaler object.
+func SetDefaults_ClusterAutoscaler(obj *ClusterAutoscaler) {
```
Comment (vpnachev) [inline pkg/apis/core/v1beta1/defaults.go:325]: I have no hard opinion, but defaulting already existing fields should be done in separate PR.

**Review by rfranzke — COMMENTED** [review]
Nice PR!

File: pkg/apis/core/v1beta1/defaults.go (hunk 2 above)
Comment (prashanth26) [reply pkg/apis/core/v1beta1/defaults.go:325]: It's a good point @vpnachev. I will keep it in mind for future reference. However, for this change, I would like to go with this single PR/commit in the interest of time. I hope it's alright.

File: pkg/apis/core/validation/shoot.go (hunk 1 above)
Comment (prashanth26) [reply pkg/apis/core/validation/shoot.go:805]: @vpnachev - I have tried to address this with https://github.com/gardener/gardener/pull/4508/commits/a4cfbbf8b07d69c6d948fd78b6a52ced55a069bb. PTAL.

File: pkg/apis/core/v1alpha1/defaults.go [hunk 3]
```diff
@@ -321,6 +321,34 @@ func SetDefaults_Worker(obj *Worker) {
 	}
 }
 
+// SetDefaults_ClusterAutoscaler sets default values for ClusterAutoscaler object.
+func SetDefaults_ClusterAutoscaler(obj *ClusterAutoscaler) {
+	if obj.ScaleDownDelayAfterAdd == nil {
+		obj.ScaleDownDelayAfterAdd = &metav1.Duration{Duration: 1 * time.Hour}
+	}
+	if obj.ScaleDownDelayAfterDelete == nil {
+		obj.ScaleDownDelayAfterDelete = &metav1.Duration{Duration: 0}
```
Comment (vpnachev) [inline pkg/apis/core/v1alpha1/defaults.go:330]: In the documentation you have 
> ... defaults to scanInterval (defaults to ScanInterval).

File: pkg/apis/core/v1alpha1/defaults.go [hunk 4]
```diff
@@ -321,6 +321,34 @@ func SetDefaults_Worker(obj *Worker) {
...
+	}
+	if obj.ScaleDownDelayAfterFailure == nil {
+		obj.ScaleDownDelayAfterFailure = &metav1.Duration{Duration: 3 * time.Minute}
+	}
+	if obj.ScaleDownUnneededTime == nil {
+		obj.ScaleDownUnneededTime = &metav1.Duration{Duration: 30 * time.Minute}
+	}
+	if obj.ScaleDownUtilizationThreshold == nil {
+		obj.ScaleDownUtilizationThreshold = pointer.Float64(0.5)
```
Comment (vpnachev) [inline pkg/apis/core/v1alpha1/defaults.go:339]: For consistency, add the default value in the documentation.

File: pkg/apis/core/v1alpha1/defaults_test.go [hunk 5]
```diff
@@ -532,6 +532,26 @@ var _ = Describe("Defaults", func() {
...
 	})
 
+	Describe("#SetDefaults_ClusterAutoscaler", func() {
+		var obj *ClusterAutoscaler
+
+		BeforeEach(func() {
+			obj = &ClusterAutoscaler{}
+		})
+
+		It("should correctly default the ClusterAutoscaler", func() {
```
Comment (vpnachev) [inline pkg/apis/core/v1alpha1/defaults_test.go:546]: Please, add a test case where already set values are not overwritten during defaulting.

File: pkg/apis/core/validation/shoot.go [hunk 6]
```diff
@@ -793,6 +799,13 @@ func ValidateClusterAutoscaler(autoScaler core.ClusterAutoscaler, fldPath *field
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("scaleDownUtilizationThreshold"), *threshold, "can not be greater than 1.0"))
 		}
 	}
+	if maxNodeProvisionTime := autoScaler.MaxNodeProvisionTime; maxNodeProvisionTime != nil && maxNodeProvisionTime.Duration < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("maxNodeProvisionTime"), *maxNodeProvisionTime, "can not be negative"))
+	}
```
Comment (vpnachev) [inline pkg/apis/core/validation/shoot.go:804]:
```suggestion
	}

```

File: pkg/apis/core/validation/shoot.go (hunk 1 above)
Comment (vpnachev) [reply pkg/apis/core/validation/shoot.go:805]: Thanks!

File: pkg/operation/botanist/component/clusterautoscaler/cluster_autoscaler.go [hunk 7]
```diff
@@ -339,8 +339,14 @@ func (c *clusterAutoscaler) emptyManagedResourceSecret() *corev1.Secret {
...
-		scaleDownUnneededTime  = metav1.Duration{Duration: 30 * time.Minute}
-		scaleDownDelayAfterAdd = metav1.Duration{Duration: time.Hour}
+		expander                      = "least-waste"
+		scaleDownUtilizationThreshold = 0.5
+		maxNodeProvisionTime          = metav1.Duration{Duration: 20 * time.Minute}
+		scaleDownUnneededTime         = metav1.Duration{Duration: 30 * time.Minute}
+		scaleDownDelayAfterAdd        = metav1.Duration{Duration: time.Hour}
+		scaleDownDelayAfterFailure    = metav1.Duration{Duration: 3 * time.Minute}
+		scaleDownDelayAfterDelete     = metav1.Duration{Duration: 0}
+		scanInterval                  = metav1.Duration{Duration: 10 * time.Second}
```
Comment (vpnachev) [inline pkg/operation/botanist/component/clusterautoscaler/cluster_autoscaler.go]: I am afraid these are the actual default values and they can easily changed without doing the same in the API. I would propose to always use the one from the API as all of them are now defaulted.

File: pkg/apis/core/types_shoot.go [hunk 8]
```diff
@@ -325,8 +325,31 @@ type ClusterAutoscaler struct {
 	ScaleDownUtilizationThreshold *float64
 	// ScanInterval how often cluster is reevaluated for scale up or down (default: 10 secs).
 	ScanInterval *metav1.Duration
+	// Expander defines the algorithm to use during scale up (default: least-waste).
+	// See: https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/FAQ.md#what-are-expanders.
+	Expander *string
```
Comment (vpnachev) [inline pkg/apis/core/types_shoot.go]: You can use something like 
```golang
type ExpanderMode string
const (
	ClusterAutoscalerExpanderLeastWaste ExpanderMode = "least-waste"
	ClusterAutoscalerExpanderMostPods ExpanderMode = "most-pods"
...

	Expander *ExpanderMode
...	
```

**Review by vpnachev — COMMENTED** [review]
Thanks!

Just some nits.

File: pkg/apis/core/v1alpha1/defaults.go (hunk 3 above)
Comment (prashanth26) [reply pkg/apis/core/v1alpha1/defaults.go:330]: This is more related to CA's flag behavior where setting it to `0` defaults it to ScanIntervel. Refer [here](https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/main.go#L93-L94)

File: pkg/apis/core/v1alpha1/defaults.go (hunk 4 above)
Comment (prashanth26) [reply pkg/apis/core/v1alpha1/defaults.go:339]: Done [here](https://github.com/gardener/gardener/pull/4508/commits/ed626583966d98e816e697c1b4780373f0a3739d#diff-63c59db8a50bf6d0d486e8ae5b3ec2b334d192fe2f1bbe8d27c0c6742eeb51c3R2629)

File: pkg/apis/core/v1alpha1/defaults_test.go (hunk 5 above)
Comment (prashanth26) [reply pkg/apis/core/v1alpha1/defaults_test.go:546]: Done [here](https://github.com/gardener/gardener/pull/4508/commits/ed626583966d98e816e697c1b4780373f0a3739d#diff-d471b1b001af7d26129042984376f77d89fc79f249ad2f536ae7afcf64cc2e86R558-R578)

File: pkg/apis/core/types_shoot.go (hunk 8 above)
Comment (prashanth26) [reply pkg/apis/core/types_shoot.go]: Done [here](https://github.com/gardener/gardener/pull/4508/commits/ed626583966d98e816e697c1b4780373f0a3739d#diff-2a20577716c98d94c2d517871253ed0e37c0051d9caf6bc099f1edd1429c0ab8R335-R337).

File: pkg/operation/botanist/component/clusterautoscaler/cluster_autoscaler.go (hunk 7 above)
Comment (prashanth26) [reply pkg/operation/botanist/component/clusterautoscaler/cluster_autoscaler.go]: From my local tests, I see that the defaulting isn't done for existing shoot objects. So not sure what would be the best way for these objects. As an improvement I have updated this logic to [make use of the defaulting method here](https://github.com/gardener/gardener/blob/ed626583966d98e816e697c1b4780373f0a3739d/pkg/operation/botanist/component/clusterautoscaler/cluster_autoscaler.go#L339-L376). I am guessing there should be a better way to default. Please do suggest if there is.

File: pkg/apis/core/v1alpha1/defaults.go (hunk 3 above)
Comment (vpnachev) [reply pkg/apis/core/v1alpha1/defaults.go:330]: But this is something internal to the cluster autoscaler and Gardener does not need to know about it. So, please set in the documentation that the default value is `0`.

File: pkg/apis/core/v1alpha1/defaults_test.go [hunk 9]
```diff
@@ -532,6 +532,53 @@ var _ = Describe("Defaults", func() {
...
+			Expect(obj.ScaleDownDelayAfterAdd).To(PointTo(Equal(metav1.Duration{Duration: 1 * time.Hour})))
+			Expect(obj.ScaleDownDelayAfterDelete).To(PointTo(Equal(metav1.Duration{Duration: 0})))
+			Expect(obj.ScaleDownDelayAfterFailure).To(PointTo(Equal(metav1.Duration{Duration: 3 * time.Minute})))
+			Expect(obj.ScaleDownUnneededTime).To(PointTo(Equal(metav1.Duration{Duration: 30 * time.Minute})))
+			Expect(obj.ScanInterval).To(PointTo(Equal(metav1.Duration{Duration: 10 * time.Second})))
+			Expect(obj.MaxNodeProvisionTime).To(PointTo(Equal(metav1.Duration{Duration: 20 * time.Minute})))
+			Expect(obj.Expander).To(PointTo(Equal(expanderLeastWaste)))
+		})
+
+		It("should correctly override values the ClusterAutoscaler on setting them", func() {
```
Comment (vpnachev) [inline pkg/apis/core/v1alpha1/defaults_test.go]:
```suggestion
		It("should not default explicitly set fields", func() {
```

**Review by vpnachev — COMMENTED** [review]
Just one nit apart from the default value for `ScaleDownDelayAfterDelete`.

File: pkg/operation/botanist/component/clusterautoscaler/cluster_autoscaler.go (hunk 7 above)
Comment (vpnachev) [reply pkg/operation/botanist/component/clusterautoscaler/cluster_autoscaler.go]: I don't have better idea. I just checked other components and they also duplicate the default values, but I really think this is error prone and the approach that you now implemented seems better to me.


The defaulting does not happen because of
```golang
	if in.Spec.Kubernetes.ClusterAutoscaler != nil {
		SetDefaults_ClusterAutoscaler(in.Spec.Kubernetes.ClusterAutoscaler)
	}
```

in the `zz_generated.defaults.go`. You see, it expects the `ClusterAutoscaler` to be not nil in order to default the values, but with 
https://github.com/gardener/gardener/blob/6cdb8dc690c2464558c228105852eb3ac7c2c05a/pkg/apis/core/v1beta1/types_shoot.go#L353-L354

it is `nil` when skipped.

File: pkg/apis/core/v1alpha1/defaults.go (hunk 3 above)
Comment (prashanth26) [reply pkg/apis/core/v1alpha1/defaults.go:330]: Okay understood. Apologies for the confusion. I have [fixed the docs](https://github.com/gardener/gardener/pull/4508/commits/eaed0c70e3822c2a43c0c497bbfd8ab5ea4af716) and combined all the recent suggestions in the same commit.

**Comment by prashanth26** [comment]
Tried to address all recent changes with [this commit](https://github.com/gardener/gardener/pull/4508/commits/eaed0c70e3822c2a43c0c497bbfd8ab5ea4af716).

File: pkg/operation/botanist/component/clusterautoscaler/cluster_autoscaler.go (hunk 7 above)
Comment (prashanth26) [reply pkg/operation/botanist/component/clusterautoscaler/cluster_autoscaler.go]: Oh okay. Thanks for the detailed explanation. I knew something did not default properly, however wasn't sure where. This explanation definitely helps. Thanks.

**Review by vpnachev — APPROVED** [review]
/lgtm

**Review by rfranzke — APPROVED** [review]

File: pkg/apis/core/types_shoot.go [hunk 10]
```diff
@@ -315,18 +315,44 @@ type Kubernetes struct {
...
+// ExpanderMode is type used for Expander values
+type ExpanderMode string
+
+const (
+	// ClusterAutoscalerExpanderLeastWaste selects the node group that will have the least idle CPU (if tied, unused memory) after scale-up.
+	// This is useful when you have different classes of nodes, for example, high CPU or high memory nodes, and
+	// only want to expand those when there are pending pods that need a lot of those resources.
+	// This is the default value.
+	ClusterAutoscalerExpanderLeastWaste ExpanderMode = "least-waste"
+	// ClusterAutoscalerExpanderRandom selects the node group that would be able to schedule the most pods when scaling up.
```
Comment (ialidzhikov) [inline pkg/apis/core/types_shoot.go:344]:
```suggestion
	// ClusterAutoscalerExpanderMostPods selects the node group that would be able to schedule the most pods when scaling up.
```

File: pkg/apis/core/types_shoot.go [hunk 11]
```diff
@@ -315,18 +315,44 @@ type Kubernetes struct {
...
+	// This is useful when you are using nodeSelector to make sure certain pods land on certain nodes.
+	// Note that this won't cause the autoscaler to select bigger nodes vs. smaller, as it can add multiple smaller nodes at once.
+	ClusterAutoscalerExpanderMostPods ExpanderMode = "most-pods"
+	// ClusterAutoscalerExpanderRandom selects the node group that has the highest priority assigned by the user. For configurations,
```
Comment (ialidzhikov) [inline pkg/apis/core/types_shoot.go:348]:
```suggestion
	// ClusterAutoscalerExpanderPriority selects the node group that has the highest priority assigned by the user. For configurations,
```

File: pkg/apis/core/v1alpha1/types_shoot.go [hunk 12]
```diff
@@ -381,14 +381,42 @@ type ClusterAutoscaler struct {
...
+// ExpanderMode is type used for Expander values
+type ExpanderMode string
+
+const (
+	// ClusterAutoscalerExpanderLeastWaste selects the node group that will have the least idle CPU (if tied, unused memory) after scale-up.
+	// This is useful when you have different classes of nodes, for example, high CPU or high memory nodes, and
+	// only want to expand those when there are pending pods that need a lot of those resources.
+	// This is the default value.
+	ClusterAutoscalerExpanderLeastWaste ExpanderMode = "least-waste"
+	// ClusterAutoscalerExpanderRandom selects the node group that would be able to schedule the most pods when scaling up.
```
Comment (ialidzhikov) [inline pkg/apis/core/v1alpha1/types_shoot.go:408]:
```suggestion
	// ClusterAutoscalerExpanderMostPods selects the node group that would be able to schedule the most pods when scaling up.
```

File: pkg/apis/core/v1alpha1/types_shoot.go [hunk 13]
```diff
@@ -381,14 +381,42 @@ type ClusterAutoscaler struct {
...
+	// This is useful when you are using nodeSelector to make sure certain pods land on certain nodes.
+	// Note that this won't cause the autoscaler to select bigger nodes vs. smaller, as it can add multiple smaller nodes at once.
+	ClusterAutoscalerExpanderMostPods ExpanderMode = "most-pods"
+	// ClusterAutoscalerExpanderRandom selects the node group that has the highest priority assigned by the user. For configurations,
```
Comment (ialidzhikov) [inline pkg/apis/core/v1alpha1/types_shoot.go:412]:
```suggestion
	// ClusterAutoscalerExpanderPriority selects the node group that has the highest priority assigned by the user. For configurations,
```

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 14]
```diff
@@ -388,14 +388,42 @@ type ClusterAutoscaler struct {
...
+// ExpanderMode is type used for Expander values
+type ExpanderMode string
+
+const (
+	// ClusterAutoscalerExpanderLeastWaste selects the node group that will have the least idle CPU (if tied, unused memory) after scale-up.
+	// This is useful when you have different classes of nodes, for example, high CPU or high memory nodes, and
+	// only want to expand those when there are pending pods that need a lot of those resources.
+	// This is the default value.
+	ClusterAutoscalerExpanderLeastWaste ExpanderMode = "least-waste"
+	// ClusterAutoscalerExpanderRandom selects the node group that would be able to schedule the most pods when scaling up.
```
Comment (ialidzhikov) [inline pkg/apis/core/v1beta1/types_shoot.go:415]:
```suggestion
	// ClusterAutoscalerExpanderMostPods selects the node group that would be able to schedule the most pods when scaling up.
```

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 15]
```diff
@@ -388,14 +388,42 @@ type ClusterAutoscaler struct {
...
+	// This is useful when you are using nodeSelector to make sure certain pods land on certain nodes.
+	// Note that this won't cause the autoscaler to select bigger nodes vs. smaller, as it can add multiple smaller nodes at once.
+	ClusterAutoscalerExpanderMostPods ExpanderMode = "most-pods"
+	// ClusterAutoscalerExpanderRandom selects the node group that has the highest priority assigned by the user. For configurations,
```
Comment (ialidzhikov) [inline pkg/apis/core/v1beta1/types_shoot.go:419]:
```suggestion
	// ClusterAutoscalerExpanderPriority selects the node group that has the highest priority assigned by the user. For configurations,
```
